Привет, меня зовут Люман. Сегодня я проверю твой проект.

Комментарии будут в <font color='green'>зеленой</font>, <font color='blue'>синей</font> или <font color='red'>красной</font> рамках:

<div class="alert alert-block alert-success">
<b>Успех:</b> Если все сделано отлично
</div>

<div class="alert alert-block alert-info">
<b>Совет: </b> Если можно немного улучшить
</div>

<div class="alert alert-block alert-danger">
<b>Ошибка:</b> Если требуются исправления. Работа не может быть принята с красными комментариями.
</div>

### <font color='orange'>Общее впечатление</font>
* Большое спасибо за проделанную работу, видно, что приложено много усилий!
* Понравилось подробное составление выводов, однако хотелось бы их наблюдать на каждом шаге.
* Я оставил некоторые советы, надеюсь они будут тебе полезны.
* Поправь пожалуйста ошибку с неверными моделями, и метриками для задачи классификации. И обрати внимание пожалуйста на дублирование кода, которые ты делаешь при тестировании модели с определнными параметрами.
* Поправь эти недочеты, и проект будет зачтен. Я уверен у тебя это легко получится!


### <font color='orange'>Общее впечатление[2]</font>
* Спасибо за то, что прислала работу и поправила все недочеты!
* Видно что ты  очень старалась. У тебя получилось выполнить верно все шаги.
* Ты достигла хорошего качества на тесте!
* Твой проект принимается. Удачи в дальнейшем!

# Рекомендация тарифов

Оператор мобильной связи «Мегалайн» выяснил: многие клиенты пользуются архивными тарифами. Они хотят построить систему, способную проанализировать поведение клиентов и предложить пользователям новый тариф: «Смарт» или «Ультра».
В нашем распоряжении данные о поведении клиентов, которые уже перешли на эти тарифы. Нужно построить модель для задачи классификации, которая выберет подходящий тариф. 

Каждый объект в наборе данных — это информация о поведении одного пользователя за месяц. Известно:
сalls — количество звонков,
minutes — суммарная длительность звонков в минутах,
messages — количество sms-сообщений,
mb_used — израсходованный интернет-трафик в Мб,
is_ultra — каким тарифом пользовался в течение месяца («Ультра» — 1, «Смарт» — 0).

Необходимо построить модель с максимально большим значением *accuracy*. 

## 1. Откройте и изучите файл

In [1]:
import pandas as pd
import numpy as np

import warnings
warnings.filterwarnings("ignore")

from sklearn.metrics import accuracy_score
from sklearn.tree import DecisionTreeRegressor
from sklearn.tree import DecisionTreeClassifier

from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.datasets import make_classification

from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error
from sklearn.dummy import DummyClassifier



<div class="alert alert-block alert-success">
<b>Успех:</b> Молодец, что коллекционируешь импорты в первой ячейке, отдельный + за то что обрабатываешь warnings.
</div>


In [2]:
behavior = pd.read_csv('/datasets/users_behavior.csv')
behavior.info()
behavior.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3214 entries, 0 to 3213
Data columns (total 5 columns):
calls       3214 non-null float64
minutes     3214 non-null float64
messages    3214 non-null float64
mb_used     3214 non-null float64
is_ultra    3214 non-null int64
dtypes: float64(4), int64(1)
memory usage: 125.7 KB


calls  minutes  messages   mb_used  is_ultra
0   40.0   311.90      83.0  19915.42         0
1   85.0   516.75      56.0  22696.96         0
2   77.0   467.66      86.0  21060.45         0
3  106.0   745.53      81.0   8437.39         1
4   66.0   418.74       1.0  14502.75         0

Получили обычный дата фрейм без пропусков и нулевых значений. Целевой признак - is_ultra: 0 - тариф Смарт, 1 - тариф ультра. 
Далее выделяем признаки - features и целевой признак - target.

In [3]:
features = behavior.drop(['is_ultra'], axis=1)
target = behavior['is_ultra']

In [4]:
print(features)
print(target)

      calls  minutes  messages   mb_used
0      40.0   311.90      83.0  19915.42
1      85.0   516.75      56.0  22696.96
2      77.0   467.66      86.0  21060.45
3     106.0   745.53      81.0   8437.39
4      66.0   418.74       1.0  14502.75
...     ...      ...       ...       ...
3209  122.0   910.98      20.0  35124.90
3210   25.0   190.36       0.0   3275.61
3211   97.0   634.44      70.0  13974.06
3212   64.0   462.32      90.0  31239.78
3213   80.0   566.09       6.0  29480.52

[3214 rows x 4 columns]
0       0
1       0
2       0
3       1
4       0
       ..
3209    1
3210    0
3211    0
3212    0
3213    1
Name: is_ultra, Length: 3214, dtype: int64


In [5]:
print(len(features))

3214


<div class="alert alert-block alert-success">
<b>Успех:</b> Первый шаг сделан хорошо, все требуемые методы использованы!
</div>

<div class="alert alert-block alert-info">
<b>Совет: </b> Хотелось бы видеть побольше выводов по проделанным шагам.
</div>

## 2. Разбейте данные на выборки

In [6]:
# Делим выборку на обучащую и валидационную
features_train, features_valid, target_train, target_valid = train_test_split(
    features, target, test_size=0.40, random_state=12345) 

Сначала смотрим на получившиеся выборки

In [7]:
features_train.describe()

calls      minutes     messages       mb_used
count  1928.000000  1928.000000  1928.000000   1928.000000
mean     62.960581   436.494959    39.213693  17007.872889
std      32.881705   232.469903    36.686684   7497.741728
min       0.000000     0.000000     0.000000      0.000000
25%      41.000000   275.525000    10.000000  12497.390000
50%      62.000000   427.935000    31.000000  16663.300000
75%      82.000000   563.802500    58.000000  21304.777500
max     198.000000  1390.220000   201.000000  44296.860000

In [8]:
target_train.describe()

count    1928.000000
mean        0.307573
std         0.461608
min         0.000000
25%         0.000000
50%         0.000000
75%         1.000000
max         1.000000
Name: is_ultra, dtype: float64

In [9]:
# Делим на тестовую и валидационную
# В дальнейших проектах от ревьюера было замечание, что мы должны делить валидационную выборку пополам, т е test_size= 0,5 должен быть в данном случае). ПРОВЕРИТЬ!!! 4.04.22
features_test, features_valid_1, target_test, target_valid_1 = train_test_split(
    features_valid,target_valid, test_size=0.20, random_state=12345) 

In [10]:
features_test.describe()

calls      minutes     messages       mb_used
count  1028.000000  1028.000000  1028.000000   1028.000000
mean     63.113813   440.795282    38.077821  17460.767685
std      33.373337   234.756958    36.234707   7648.775880
min       0.000000     0.000000     0.000000      0.000000
25%      39.000000   275.890000     9.000000  12472.822500
50%      62.000000   439.975000    29.500000  17212.355000
75%      83.000000   587.722500    57.000000  21445.547500
max     244.000000  1632.060000   224.000000  49745.730000

In [11]:
features_valid_1.describe()

calls      minutes    messages       mb_used
count  258.000000   258.000000  258.000000    258.000000
mean    63.325581   440.710116   32.124031  17692.308527
std     35.390245   249.798173   30.872780   7779.168727
min      0.000000     0.000000    0.000000      0.000000
25%     38.250000   272.497500    7.250000  12482.505000
50%     60.000000   412.010000   23.000000  17851.720000
75%     81.750000   575.792500   49.500000  21913.810000
max    203.000000  1566.450000  142.000000  45180.750000

In [12]:
target_test.describe()

count    1028.000000
mean        0.303502
std         0.459994
min         0.000000
25%         0.000000
50%         0.000000
75%         1.000000
max         1.000000
Name: is_ultra, dtype: float64

In [13]:
target_valid_1.describe()

count    258.000000
mean       0.310078
std        0.463424
min        0.000000
25%        0.000000
50%        0.000000
75%        1.000000
max        1.000000
Name: is_ultra, dtype: float64

Каких-то странных показателей, не похожих на реальные не видно.

<div class="alert alert-block alert-success">
<b>Успех:</b> Разбиение сделано абсолютно верно
</div>

<div class="alert alert-block alert-info">
<b>Совет: </b> Хорошим тоном является выводить размеры выборок после разбиений, это может быть очень полезным для предотвращения ошибок - ВОЗМОЖНО Я НЕПРАВИЛЬНО ПОНЯЛА ЧТО НАДО СДЕЛАТЬ-ПРИМЕНИЛА МЕТОД  describe.
</div>

<div class="alert alert-block alert-success">
<b>Успех[2]:</b> Describe тоже сойдет, но достаточно использовать метод shape.
</div>

## 3. Исследуйте модели

Перед нами стоит задача классификации - необходимо определить наиболее подходящий тариф.
При построении моделей для решения задач классификации используют следующие модели:
"DecisionTreeClassifier", "RandomForestClassifier" и "LogisticRegression".

In [14]:
# Исследуем модель Дерево решений

model = DecisionTreeClassifier(random_state=12345)

model.fit(features, target)

DecisionTreeClassifier(class_weight=None, criterion='gini', max_depth=None,
                       max_features=None, max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, presort=False,
                       random_state=12345, splitter='best')

In [15]:
# DecisionTreeClassifier
best_model = None
best_result = 0
for depth in range(1, 6):
    model = DecisionTreeClassifier(random_state=12345, max_depth=depth)
    model.fit(features_train, target_train)
    predictions_valid_1 = model.predict(features_valid_1)
    result = accuracy_score(target_valid_1, predictions_valid_1)
    if result > best_result:
        best_model = model
        best_result = result
print("Accuracy наилучшей модели на валидационной выборке:", best_result)    
    

Accuracy наилучшей модели на валидационной выборке: 0.8062015503875969


Вывод: accuracy_score (точность модели) на валидационной выборке, полученной на модели обученной на обучающей выборке, составляет 80,6%.

In [16]:
#Случайный лес
best_model = None
best_result = 0
for est in range(1, 11):
    model = RandomForestClassifier(random_state=12345, n_estimators=est) 
    model.fit(features_train, target_train) # обучаем модель на тренировочной выборке
    result = model.score(features_valid_1, target_valid_1) 
    if result > best_result:
        best_model = model
        best_result = result
print('Accuracy наилучшей модели на валидационной выборке', best_result)

Accuracy наилучшей модели на валидационной выборке 0.7790697674418605


In [17]:
# Логистическая регрессия
model = LogisticRegression(random_state=12345)
model.fit(features_train, target_train) 
result = model.score(features_valid_1, target_valid_1) 

print("Accuracy модели логистической регрессии на валидационной выборке:", result)

Accuracy модели логистической регрессии на валидационной выборке: 0.7441860465116279


Вывод: Наиболее высокие показатели Accuracy на валидационной выборке показала модель DecisionTreeClassifier = 0.8062, наименьший показатель наилучшей модели показала модель RandomForestClassifier = 0.7015, модель  Логистическая регрессия показала значение для наилучшей модели 0.7441.
Accuracy - показывает количество правильно проставленных меток класса (истинно положительных и истинно отрицательных) от общего количества данных. 
Для исследования качества разных моделей, меняя гиперпараметры, будем использовать модель показавшую наибольшее значение Accuracy - DecisionTreeClassifier.


## Исследуйте качество разных моделей, меняя гиперпараметры

Наиболее важным гиперпараметром является величина max_depth.Изменим данный гиперпараметр в цикле путем перебора, параметр random_state должен оставаться постоянным.

In [18]:
# модель DecisionTreeClassifier смена гиперпараметра max_depth
for depth in range(1, 10):
    model = DecisionTreeClassifier(random_state=12345, max_depth=depth)

    model.fit(features_train, target_train)

    predictions_valid_1 = model.predict(features_valid_1)
  
    print("max_depth =", depth, ": ", end='')
    print(accuracy_score(target_valid_1, predictions_valid_1))
    

max_depth = 1 : 0.748062015503876
max_depth = 2 : 0.7945736434108527
max_depth = 3 : 0.8023255813953488
max_depth = 4 : 0.7868217054263565
max_depth = 5 : 0.8062015503875969
max_depth = 6 : 0.7790697674418605
max_depth = 7 : 0.7984496124031008
max_depth = 8 : 0.7868217054263565
max_depth = 9 : 0.7790697674418605


Вывод по исследованию гиперпараметров модели "DecisionTreeClassifier" :
При переборе гиперпараметра max_depth модели  "DecisionTreeClassifier" в цикле от 1 до 9 максимальная точность модели была получена при максимальной глубине дерева = 5. Далее точность модели начала снижаться.

In [19]:
#Случайный лес-смена гиперпараметра n_estimators и max_depth
best_model = None
best_result = 0
for est in range(1, 11):
    for depth in range (1, 10):
        model = RandomForestClassifier(random_state=12345, n_estimators=est, max_depth=depth) 
        model.fit(features_train, target_train) # обучаем модель на тренировочной выборке
        result = model.score(features_valid_1, target_valid_1) 
        if result > best_result:
            best_model = model
            best_result = result
            best_est = est
            best_depth = depth
             
print('Accuracy наилучшей модели на валидационной выборке',  best_result, 'максимальная глубина', best_depth, 'Количество деревьев', best_est)

Accuracy наилучшей модели на валидационной выборке 0.8255813953488372 максимальная глубина 9 Количество деревьев 1


Вывод по исследованию гиперпараметров модели "RandomForestClassifier":
При переборе гиперпараметров n_estimators модели "RandomForestClassifier" в цикле от 1 до 10 и параметра  max_depth от 1-10, максимальная точность модели была получена при количестве деревьев 1 и максимальной глубине дерева 9. Accuracy = 0.8255813953488372.

In [20]:
# применим гиперпараметр Кол-во деревьев 5(больше идеального) и глубина 10 (больше идеального)
model = RandomForestClassifier(random_state=12345, n_estimators=5, max_depth=10) 
model.fit(features_train, target_train) 
result = model.score(features_valid_1, target_valid_1) 
print('Accuracy', result)

Accuracy 0.8178294573643411


In [21]:
# применим гиперпараметр Кол-во деревьев 3(больше идеального) и глубина 5 (меньше идеального)
model = RandomForestClassifier(random_state=12345, n_estimators=3, max_depth=5) 
model.fit(features_train, target_train) 
result = model.score(features_valid_1, target_valid_1) 
print('Accuracy', result)

Accuracy 0.8023255813953488


Вывод:
при изменении параметров Количество деревьев и максимальнеая глубина (как в сторону увеличения так и в сторону уменьшения) была выявлена тенденция к снижению точности нашей модели.


<div class="alert alert-block alert-danger">
<b>Ошибка устраненная:</b> Хочу отметить некоторые недочеты: <br>
    1) Ты в двойном цикле используешь одну и ту же модель 10 раз, вместо 1 раза. (Циклы не нужны если ты хочешь проверить конкретные параметры для модели) <br>
    2) Ты уже проделывал эти вычисления перебором параметров, ты мог выводить с помощью print эти значения. И тебе не нужно было бы еще раз проделывать тоже самое.
</div>

<div class="alert alert-block alert-danger">
<b>Ошибка устраненная:</b> RMSE метрика для задач регресии, использовать ее для задачи классификации не совсем корректно. <br>
Также выбраны неверно модели для тестирования. <br> Нужно использовать  RandomForestClassifier вместо RandomForestRegressor, DecisionTreeClassifier вместо DecisionTreeRegressor и т.п .<br> Наверное тебя спутала LogisticRegression, которая является моделью для решения задач классификации. 
    
[Откуда у нее такое название можно прочитать здесь](https://habr.com/ru/company/io/blog/265007/)
</div>

<div class="alert alert-block alert-info">
<b>Совет: </b> Хорошо было бы посчитать, хотя бы для одной модели  больше 2 гиперпараметров.
    
[Узнать какие параметры принимает модель можно в официальной документации.
    ](https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.RandomForestClassifier.html)
</div>

<div class="alert alert-block alert-success">
<b>Успех[2]:</b> Молодец! Ты поправила все недочеты! Отлично, что ты использовала для подбора параметров только валидационную выборку.
</div>

## 4. Проверьте модель на тестовой выборке

In [22]:
# Проверяем модель "Дерево решений"
model = DecisionTreeClassifier(random_state=12345)

model.fit(features_test, target_test)

DecisionTreeClassifier(class_weight=None, criterion='gini', max_depth=None,
                       max_features=None, max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, presort=False,
                       random_state=12345, splitter='best')

In [23]:
# DecisionTreeClassifier
best_model = None
best_result = 0
for depth in range(1, 6):
    model = DecisionTreeClassifier(random_state=12345, max_depth=depth)
    model.fit(features_train, target_train)
    predictions_test = model.predict(features_test)
    result = accuracy_score(target_test, predictions_test)
    if result > best_result:
        best_model = model
        best_result = result
print("Accuracy наилучшей модели на тестовой выборке:", best_result)    
    

Accuracy наилучшей модели на тестовой выборке: 0.7772373540856031


Accuracy наилучшей модели на тестовой выборке ниже, чем на валидационной (0,806). Это может говорить о том, что в качестве тестовой выборки лучше взять валиадационную выборку(нежели тестовую), при подборе рабочей модели лучше ориентироваться на ее данные. Также, размер нашей выборки достаточно маленький, это также может влиять на точность.

In [24]:
#Случайный лес
best_model = None
best_result = 0
for est in range(1, 11):
    model = RandomForestClassifier(random_state=12345, n_estimators=est) 
    model.fit(features_train, target_train) # обучаем модель на тренировочной выборке
    result = model.score(features_test, target_test) 
    if result > best_result:
        best_model = model
        best_result = result
print('Accuracy наилучшей модели на тестовой выборке', best_result)

Accuracy наилучшей модели на тестовой выборке 0.7859922178988327


Вывод: Для модели "Случайный лес"  Accuracy наилучшей модели на тестовой выборке 0.7859 больше, чем Accuracy наилучшей модели на валидационной выборке = 0.7790.Это говорит о более высокой точности предсказания модели на тестовой выборке.

In [25]:
# Логистическая регрессия
model = LogisticRegression(random_state=12345)
model.fit(features_train, target_train) 
result = model.score(features_test, target_test) 

print("Accuracy наилучшей модели логистической регрессии на тестовой выборке:", result)

Accuracy наилучшей модели логистической регрессии на тестовой выборке: 0.7509727626459144


Вывод: Для модели "Логистическая регрессия"  Accuracy наилучшей модели на тестовой выборке 0.7509 больше, чем Accuracy наилучшей модели на валидационной выборке = 0.7441. Это говорит о более высокой точности предсказания модели на тестовой выборке.

In [26]:
print('Accuracy наилучшей модели на тестовой выборке "DecisionTreeClassifier":',  0.7772373540856031)  
print('Accuracy наилучшей модели на тестовой выборке "RandomForestClassifier:', 0.7859922178988327)
print("Accuracy наилучшей модели логистической регрессии на тестовой выборке:", 0.7509727626459144)

Accuracy наилучшей модели на тестовой выборке "DecisionTreeClassifier": 0.7772373540856031
Accuracy наилучшей модели на тестовой выборке "RandomForestClassifier: 0.7859922178988327
Accuracy наилучшей модели логистической регрессии на тестовой выборке: 0.7509727626459144


Вывод:
Анализ моделей, обученных на одной тренировочной выборке и протестированных на одной и тойже тестовой выборке, показал, что точность двух моделей "Случайный лес" и "Логистическая регрессия" немного больше при проверке их на тестовой выборке, в то время как точность наилучшей модели "Дерево решений" выше на валидационной выборке.
Наибольшаяточность модели на тестовой выборке показала модель RandomForestClassifier - 0.7859922178988327.


<div class="alert alert-block alert-success">
<b>Успех[2]:</b> Молодец, тебе удалось достичь хорошего качества!
</div>

## 5. (бонус) Проверьте модели на адекватность

Для проверки модели на адекватность, проверим ее с помощью DummyClassifier. Этот «классификатор» прогнозирует значения, например, наиболее частотные. И если алгоритм нашей модели лучше, чем рандомный ответ, то модель прошла проверку на адекватность.

In [27]:
dummy_clf = DummyClassifier(strategy="most_frequent")
dummy_clf.fit(features_train, target_train)
DummyClassifier(strategy='most_frequent')

print(dummy_clf.predict(features_test))

dummy_clf.score(features_test, target_test)

[0 0 0 ... 0 0 0]


0.6964980544747081

Результат DummyClassifier = 0.6964, в то время как Accuracy наилучшей модели на тестовой выборке: 0.7859922178988327. Получается, что точнсть нашей модели выше,следовательно модель прошла оценку на адекватность.

<div class="alert alert-block alert-success">
<b>Успех:</b> Проверка на адекватность выполнена корректно.
</div> 

## Чек-лист готовности проекта

Поставьте 'x' в выполненных пунктах. Далее нажмите Shift+Enter.

- [x] Jupyter Notebook открыт
- [x] Весь код исполняется без ошибок
- [x] Ячейки с кодом расположены в порядке исполнения
- [x] Выполнено задание 1: данные загружены и изучены
- [x] Выполнено задание 2: данные разбиты на три выборки
- [x] Выполнено задание 3: проведено исследование моделей
    - [x] Рассмотрено больше одной модели
    - [x] Рассмотрено хотя бы 3 значения гипепараметров для какой-нибудь модели
    - [x] Написаны выводы по результатам исследования
- [x] Выполнено задание 3: Проведено тестирование
- [x] Удалось достичь accuracy не меньше 0.75
